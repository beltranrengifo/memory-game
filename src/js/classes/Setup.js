import Modal from '@/js/classes/Modal'
import axios from 'axios'
import { getUser, getLocation, getGameLevel } from '@/js/utils/modal-templates'
import { geolocationUrl } from '@/js/utils/locationiq'

export default class Setup {
  constructor () {
    this.modal = new Modal()
    this.modal.showModal()
    this.user = null
    this.userLocation = null
    this.level = null
  }

  getUser () {
    return new Promise(resolve => {
      this.modal.setContent(getUser())
      const button = this.modal.modalBody.querySelector('#user-modal .button')
      button.addEventListener('click', () => {
        const value = this.modal.modalBody.querySelector('#input-user').value
        if (value) {
          this.user = this.modal.modalBody.querySelector('#input-user').value
          this.modal.deleteContent()
          resolve(this.user)
        } else {
          button.innerText = 'Please, type your name'
        }
      })
    })
  }

  getUserLocation () {
    return new Promise(resolve => {
      this.modal.setContent(getLocation())
      const buttons = this.modal.modalBody.querySelectorAll('#location-modal .button')
      Array.from(buttons).forEach(button => {
        button.addEventListener('click', event => {
          if (event.target.value) {
            this.modal.showLoader()
            navigator.geolocation.getCurrentPosition(
              position => {
                axios.get(geolocationUrl(position.coords.latitude, position.coords.longitude))
                  .then(data => {
                    this.userLocation = {
                      latitude: position.coords.latitude,
                      longitude: position.coords.longitude,
                      geocode: data.data
                    }
                    this.modal.hideLoader()
                    resolve(this.userLocation)
                  })
              },
              error => {
                this.userLocation = false
                console.log(error)
                resolve(this.userLocation)
              }
            )
          } else {
            this.userLocation = false
          }
        })
      })
    })
  }

  getGameLevel () {
    return new Promise(resolve => {
      this.modal.setContent(getGameLevel(this.user, this.userLocation && this.userLocation.geocode.address.city))
      const buttons = this.modal.modalBody.querySelectorAll('#level-modal button.levels__level')
      Array.from(buttons).forEach(button => {
        button.addEventListener('click', event => {
          this.level = parseInt(event.target.value)
          setTimeout(() => this.modal.destroyModal(), 100)
          resolve(this.getGameConfig())
        })
      })
    })
  }

  getGameConfig () {
    return {
      user: this.user,
      userLocation: this.userLocation,
      level: this.level
    }
  }
}
