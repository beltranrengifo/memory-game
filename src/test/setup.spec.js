import Setup from '@/js/classes/Setup'

describe('Setup Constructor', () => {
  let setup
  beforeEach(() => {
    setup = new Setup()
    setup.user = 'Pepe'
    setup.userLocation = {
      lat: 123,
      long: 123
    }
    setup.level = 5
  })

  test('Setup class should be a function', () => {
    expect(typeof Setup).toBe('function')
  })

  test('Setup should be defined after creation', () => {
    expect(setup).toBeDefined()
  })

  test('Setup should create modal', () => {
    expect(setup.modal).toBeDefined()
  })
})

describe('getUser method', () => {
  let setup
  beforeEach(() => {
    setup = new Setup()
    setup.user = 'Pepe'
    setup.userLocation = {
      lat: 123,
      long: 123
    }
    setup.level = 5
  })

  test('Setup getUser should be fn', () => {
    expect(typeof setup.getUser).toBe('function')
  })

  test('Setup getUser should set the value of the user asynchronous', () => {
    // eslint-disable-next-line jest/valid-expect-in-promise
    setup.getUser().then(user => {
      expect(user).toBe('Luis')
    })
    setup.modal.modalBody.querySelector('#input-user').value = 'Luis'
    setup.modal.modalBody.querySelector('#user-modal .button').click()
  })
})

describe('getUserLocation method', () => {
  let setup
  beforeEach(() => {
    setup = new Setup()
  })

  test('Setup getUserLocation should be fn', () => {
    expect(typeof setup.getUserLocation).toBe('function')
  })
})

describe('getGameLevel method', () => {
  let setup
  beforeEach(() => {
    setup = new Setup()
  })

  test('Setup getGameLevel should be fn', () => {
    expect(typeof setup.getGameLevel).toBe('function')
  })
})

describe('getGameConfig method', () => {
  let setup
  beforeEach(() => {
    setup = new Setup()
  })

  test('Setup getGameConfig should be fn', () => {
    expect(typeof setup.getGameConfig).toBe('function')
  })

  test('Setup getGameConfig should return an obj', () => {
    const a = setup.getGameConfig()
    expect(typeof a).toBe('object')
  })

  test('Setup getGameConfig should match snapshot', () => {
    const a = setup.getGameConfig()
    expect(a).toMatchSnapshot()
  })
})
