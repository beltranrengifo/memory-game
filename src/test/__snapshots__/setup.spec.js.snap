// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGameConfig method Setup getGameConfig should match snapshot 1`] = `
Object {
  "level": null,
  "user": null,
  "userLocation": null,
}
`;
